### ISSUE TITLE
Bad file  descriptor error when restart witness node

### ISSUE THREAD

**Issue by cwyyprog** [bitshares/bitshares-core#575]
I compiled a binary from the lastest master source, with the https://github.com/bitshares/bitshares-fc/pull/8 to update the openssl to 1.1, as the websocketpp 0.6.0/0.7.0 doesn't compile with openssl 1.1, I updated the submdule to develop branch at first, then the whole project compiles successfully.
When testing the new compiled witness_node, it runs successfully without rpc server after replaying the whole blocks in the witness_node_data_dir. it failed to start up the websocket rpc server , due to the following error: 
```
3334890ms th_a       application.cpp:443           startup              ] 13 N5boost16exception_detail10clone_implINS0_19error_info_injectorINS_6system12system_errorEEEEE: set_option: Bad file descriptor
set_option: Bad file descriptor: 
    {"what":"set_option: Bad file descriptor"}
    th_a  application.cpp:284 reset_websocket_server
3334890ms th_a       application.cpp:1024          startup              ] 13 N5boost16exception_detail10clone_implINS0_19error_info_injectorINS_6system12system_errorEEEEE: set_option: Bad file descriptor
set_option: Bad file descriptor: 
    {"what":"set_option: Bad file descriptor"}
    th_a  application.cpp:284 reset_websocket_server
rethrow
    {}
    th_a  application.cpp:443 startup
3334891ms th_a       main.cpp:262                  main                 ] Exiting with error:
13 N5boost16exception_detail10clone_implINS0_19error_info_injectorINS_6system12system_errorEEEEE: set_option: Bad file descriptor
set_option: Bad file descriptor: 
    {"what":"set_option: Bad file descriptor"}
    th_a  application.cpp:284 reset_websocket_server
rethrow
    {}
    th_a  application.cpp:443 startup
3334921ms th_a       db_management.cpp:191         close                ] Rewinding from 23444623 to 23444605

```

**Comment by pmconrad** [comment]
When you say "latest source", which branch are you on? What rpc-endpoint did you set?

**Comment by abitmore** [comment]
Earlier logs would be helpful.
Command line parameters and settings in `config.ini` would be helpful as well.

**Comment by cwyyprog** [comment]
The branch is master, rpc-endpoint is 127.0.0.1:8090, 
no command line parameters, the config.ini is
```
# Endpoint for P2P node to listen on
# p2p-endpoint = 

# P2P nodes to connect to on startup (may specify multiple times)
# seed-node = 

# JSON array of P2P nodes to connect to on startup
# seed-nodes = 

# Pairs of [BLOCK_NUM,BLOCK_ID] that should be enforced as checkpoints.
# checkpoint = 

# Endpoint for websocket RPC to listen on
rpc-endpoint = 127.0.0.1:8090 

# Endpoint for TLS websocket RPC to listen on
# rpc-tls-endpoint = 

# The TLS certificate file for this server
# server-pem = 

# Password for this certificate
# server-pem-password = 

# File to read Genesis State from
# genesis-json = 

# Block signing key to use for init witnesses, overrides genesis file
# dbg-init-key = 

# JSON file specifying API permissions
# api-access = 

# Space-separated list of plugins to activate
# plugins = 

# Enable block production, even if the chain is stale.
enable-stale-production = false

# Percent of witnesses (0-99) that must be participating in order to produce blocks
required-participation = false

# ID of witness controlled by this node (e.g. "1.6.5", quotes are required, may specify multiple times)
# witness-id = 

# Tuple of [PublicKey, WIF private key] (may specify multiple times)
private-key = ["BTS6MRyAjQq8ud7hVNYcfnVPJqcVpscN5So8BhtHuGYqET5GDW5CV","5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3"]

# Tuple of [PublicKey, WIF private key] (may specify multiple times)
debug-private-key = ["BTS6MRyAjQq8ud7hVNYcfnVPJqcVpscN5So8BhtHuGYqET5GDW5CV","5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3"]

# Account ID to track history for (may specify multiple times)
# track-account = 

# Keep only those operations in memory that are related to account history tracking
partial-operations = 1

# Maximum number of operations per account will be kept in memory
max-ops-per-account = 1000

# Elastic Search database node url
# elasticsearch-node-url = 

# Number of bulk documents to index on replay(5000)
# elasticsearch-bulk-replay = 

# Number of bulk documents to index on a syncronied chain(10)
# elasticsearch-bulk-sync = 

# Log bulk events to database
# elasticsearch-logs = 

# Use visitor to index additional data(slows down the replay)
# elasticsearch-visitor = 

# Track market history by grouping orders into buckets of equal size measured in seconds specified as a JSON array of numbers
bucket-size = [60,300,900,1800,3600,14400,86400]

# How far back in time to track history for each bucket size, measured in the number of buckets (default: 1000)
history-per-size = 1000

# RPC endpoint of a trusted validating node (required)
# trusted-node = 

# Block number after which to do a snapshot
# snapshot-at-block = 

# Block time (ISO format) after which to do a snapshot
# snapshot-at-time = 

# Pathname of JSON file where to store the snapshot
# snapshot-to = 

# declare an appender named "stderr" that writes messages to the console
[log.console_appender.stderr]
stream=std_error

# declare an appender named "p2p" that writes messages to p2p.log
[log.file_appender.p2p]
filename=logs/p2p/p2p.log
# filename can be absolute or relative to this config file

# route any messages logged to the default logger to the "stderr" logger we
# declared above, if they are info level are higher
[logger.default]
level=info
appenders=stderr

# route messages sent to the "p2p" logger to the p2p appender declared above
[logger.p2p]
level=info
appenders=p2p

```

**Comment by cwyyprog** [comment]
the console outputs:
```
2181941ms th_a       witness.cpp:88                plugin_initialize    ] witness plugin:  plugin_initialize() begin
2181943ms th_a       witness.cpp:99                plugin_initialize    ] Public Key: BTS6MRyAjQq8ud7hVNYcfnVPJqcVpscN5So8BhtHuGYqET5GDW5CV
2181943ms th_a       witness.cpp:117               plugin_initialize    ] witness plugin:  plugin_initialize() end
2181943ms th_a       object_database.cpp:106       open                 ] Opening object database from /home/qx/bts/run/witness_node_data_dir/blockchain ...
2466014ms th_a       object_database.cpp:111       open                 ] Done opening object database.
2466060ms th_a       db_management.cpp:59          reindex              ] reindexing blockchain
2466060ms th_a       db_management.cpp:65          reindex              ] Replaying blocks, starting at 23526231...
2466212ms th_a       db_management.cpp:122         reindex              ] Done reindexing, elapsed time: 0.15171799999999999 sec
2466254ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 104.236.144.84:1777
2466263ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 128.199.143.47:2015
2466263ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 23.92.53.182:1776
2466264ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 192.121.166.162:1776
2466264ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 51.15.61.160:1776
2467676ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 114.92.224.87:62015
2468116ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 148.251.80.54:4243
2470352ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 51.15.42.228:50696
2470353ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 212.47.249.84:50696
2472731ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 45.77.136.107:1776
2472972ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 71.197.2.119:1776
2474019ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 104.145.234.6:1777
2474480ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 149.56.17.159:1776
2476130ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 108.61.176.106:1776
2477502ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 188.241.58.128:1776
2477963ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 84.200.84.162:1776
2478530ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 45.32.117.94:1776
2481379ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 188.241.58.128:1776
2481380ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 209.105.239.13:1776
2481380ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 149.56.17.159:1776
2481380ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 94.130.15.169:1776
2481381ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 192.151.158.106:1776
2481381ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 51.15.213.23:1776
2481381ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 71.197.2.119:1776
2481381ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 51.15.61.160:1776
2481382ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 45.77.136.107:1776
2481382ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 81.89.101.133:1776
2481382ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 23.92.53.182:1776
2481382ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 108.61.176.106:1776
2481383ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 45.32.159.93:1776
2481383ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 52.119.1.55:1776
2481383ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 45.79.174.179:1776
2481383ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 197.242.158.53:1776
2481383ms th_a       application.cpp:191           reset_p2p_node       ] Adding seed node 192.121.166.162:1776
2481406ms th_a       application.cpp:206           reset_p2p_node       ] Configured p2p node to listen on 0.0.0.0:37329
2481441ms th_a       application.cpp:281           reset_websocket_serv ] Configured websocket rpc to listen on 127.0.0.1:8090
2481444ms th_a       application.cpp:443           startup              ] 13 N5boost16exception_detail10clone_implINS0_19error_info_injectorINS_6system12system_errorEEEEE: set_option: Bad file descriptor
set_option: Bad file descriptor: 
    {"what":"set_option: Bad file descriptor"}
    th_a  application.cpp:284 reset_websocket_server
2481444ms th_a       application.cpp:1024          startup              ] 13 N5boost16exception_detail10clone_implINS0_19error_info_injectorINS_6system12system_errorEEEEE: set_option: Bad file descriptor
set_option: Bad file descriptor: 
    {"what":"set_option: Bad file descriptor"}
    th_a  application.cpp:284 reset_websocket_server
rethrow
    {}
    th_a  application.cpp:443 startup
2481446ms th_a       main.cpp:262                  main                 ] Exiting with error:
13 N5boost16exception_detail10clone_implINS0_19error_info_injectorINS_6system12system_errorEEEEE: set_option: Bad file descriptor
set_option: Bad file descriptor: 
    {"what":"set_option: Bad file descriptor"}
    th_a  application.cpp:284 reset_websocket_server
rethrow
    {}
    th_a  application.cpp:443 startup
2481491ms th_a       db_management.cpp:191         close                ] Rewinding from 23526245 to 23526230

```

**Comment by cwyyprog** [comment]
solved by checkouting the 0.7.0 release of websocketpp, ie https://github.com/zaphoyd/websocketpp/tree/0.7.0 and patching with https://github.com/LocutusOfBorg/websocketpp/commit/1dd07113f2a7489444a8990a95be42e035f8e9df according to: https://github.com/zaphoyd/websocketpp/issues/599

**Comment by abitmore** [comment]
Just a note for those who don't check edits but only new messages: OP is edited with more info, this issue is actually related to OpenSSL 1.1.
